dash/include/dash/algorithm/Sort.h

```diff
@@ -0,0 +1,1177 @@
+
+  //
+  // The first two elements are always 0
+  std::vector<value_t> l_nlt_nle(NLT_NLE_BLOCK * nunits, 0);
+  auto                 l_nlt_nle_data = l_nlt_nle.data();
+
+  // TODO: dash::copy does not work with BLOCKCYCLIC patterns.
+  // TODO: add unit test and create issue
+  //
+  // For this reason we explicitly use handles instead of dash::copy_async.
+  // Moreover it is faster since dash::copy makes a lot of pattern stuff and
+  // we definitely know what we are doing here.
+  std::vector<dart_handle_t> handles{};
+
+  std::transform(
+      p_unit_info.valid_remote_partitions.begin(),
+      p_unit_info.valid_remote_partitions.end(),
+      std::back_inserter(handles),
+      [it_g_nlt_nle, l_nlt_nle_data](std::size_t p) {
+        // All elements in input range are remote
+        index_t lidx      = p * NLT_NLE_BLOCK;
+        auto    src_begin = it_g_nlt_nle + lidx;
+
+        dart_handle_t handle;
+        dash::internal::get_handle(
+            src_begin.dart_gptr(),
+            l_nlt_nle_data + lidx,
+            NLT_NLE_BLOCK,
+            &handle);
+        return handle;
+      });
+
+  DASH_ASSERT_EQ(
+      handles.size(),
+      p_unit_info.valid_remote_partitions.size(),
+      "invalid size of handles vector");
+
+  auto const last_p   = p_unit_info.valid_remote_partitions.back();
+  auto       out_last = l_nlt_nle_data + last_p * NLT_NLE_BLOCK;
+
+  dash::Future<value_t*> fut_result(
+      // wait
+      [out_last, &handles]() {
+        // Wait for all get requests to complete:
+        auto _out = out_last;
+        DASH_LOG_TRACE(
+            "dash::detail::psort_validate_partitions [Future]()",
+            "  wait for",
+            handles.size(),
+            "async get request");
+        DASH_LOG_TRACE(
+            "dash::detail::psort_validate_partitions [Future]",
+            "  _out:",
+            _out);
+        if (handles.size() > 0) {
+          if (dart_waitall_local(handles.data(), handles.size()) != DART_OK) {
+            DASH_LOG_ERROR(
+                "dash::detail::psort_validate_partitions [Future]",
+                "  dart_waitall_local failed");
+            DASH_THROW(
+                dash::exception::RuntimeError,
+                "dash::detail::psort_validate_partitions [Future]: "
+                "dart_waitall_local failed");
+          }
+        }
+        else {
+          DASH_LOG_TRACE(
+              "dash::detail::psort_validate_partitions [Future]",
+              "  No pending handles");
+        }
+        DASH_LOG_TRACE(
+            "dash::detail::psort_validate_partitions [Future] >",
+            "  async requests completed, _out:",
+            _out);
+        return _out;
+      });
+
+  // copy local portion
+  index_t lidx      = myid * NLT_NLE_BLOCK;
+  auto    cpy_begin = it_g_nlt_nle + lidx;
+  auto    lbegin    = cpy_begin.local();
+  std::copy(lbegin, lbegin + NLT_NLE_BLOCK, l_nlt_nle_data + lidx);
+
+  // wait for all copies
+  fut_result.wait();
```

Comment:
Why construct a future here if you just want to call `dart_waitall_local`? ;)

Reply:
OK it is changed now. See below, it was originally "copied" from your modifications after API changes. Looks way better now.